### PR TITLE
Large patch which fixes issues with PYTHIAParticleTable, adds unit tests, please review

### PR DESCRIPTION
**Pull request by HDembinski**
Hi Anatoli,

this patch became larger than anticipated, sorry. I started by fixing the issue we discussed that some entries in the PYTHIAParticleData table are missing. While digging into the code of PYTHIAParticleData, I noticed some opportunities to make the code more readable, less redundant, and faster.

A row in the table is now stored in a class called ParticleData, which derives from a namedtuple with the fields `name`, `mass`, `ctau`, and `charge`. It uses the slots mechanism in Python, which provides faster field access and lets this class use less memory per instance. This is useful, because we have many of these items. I also considered putting the whole table into a numpy array, but (perhaps surprisingly) element lookup in a numpy array is much slower than lookup in a list.

Then I made a custom dict-like object, which allows to look up ParticleData by PDG ID or particle name. By putting this logic into the dict-like object, we don't have to repeat it in several places in methods of the PYTHIAParticleData object.

I added a lot of unit tests for PYTHIAParticleData. It is highly checked now, and it is guaranteed that the information in the xml file is consistent with the table (no missing entries, etc.).

You had a `test()` function in tables.py with some manual test code that prints to the shell. I converted this code into several example scripts, which now live in the `examples` folder. These are not unit tests, but they severe as examples on how to use the library.

Please review and let me know whether you see problems with the changes in PYTHIAParticleData.

### CODE REVIEW

File: particletools/tables.py [hunk 1]
```diff
@@ -47,120 +91,119 @@ class PYTHIAParticleData(object):
...
-                except KeyError:
-                    pass
+                    ctau = float("NaN")
+
+                self._particle_data[pdg_id] = PData(attr['name'], mass, ctau,
+                                                  charge)
+                if 'antiName' in attr:
+                    self._particle_data[-pdg_id] = PData(attr['antiName'], mass,
+                                                       ctau, -charge)
                 #Extract branching ratios and decay channels
```
Comment (afedynitch) [inline particletools/tables.py:158]: I curious what happens here if the particle does not have an anti-particle, like K0S/L. Check needed

File: particletools/tables.py [hunk 2]
```diff
@@ -26,12 +26,56 @@
 
 from abc import ABCMeta
 from tempfile import TemporaryFile
-
+from collections import namedtuple
 import six
+from six.moves import cPickle as pickle
+# speed of light in local units
+c_speed_of_light = 2.99e10
```
Comment (afedynitch) [inline particletools/tables.py]: speed of light should have more significant digits. Local units are cm/s, please add this to the comment

File: particletools/tables.py (hunk 1 above)
Comment (HDembinski) [reply particletools/tables.py:158]: If there is no anti-particle, none is added to the database. Your original code did the same, but you used try-except, while I check the xml attribute. This should be faster than a try-except and I also think it is more readable.

File: particletools/tables.py (hunk 2 above)
Comment (HDembinski) [reply particletools/tables.py]: Putting more digits is a good idea. But you know, this is your original number, I only pulled it out from inside a function.

Comment (HDembinski) [reply particletools/tables.py]: I just added a patch. It should show up automatically in the PR.

**Review by afedynitch — APPROVED** [review]
I reviewed the changes and tested the code.

It works beautifully.

**Pull request merged**